### ISSUE TITLE
supervisord service fails to start when defining programs on the class

### ISSUE THREAD

**Issue by mcrumm** [ajcrowe/puppet-supervisord#44]
I encountered a problem when trying to define programs and groups on the supervisord class.

It looks like the `create_resources` calls for groups and programs aren't being properly anchored to ensure their existence before the service attempts to start.

Here's a snippet from a puppet run, showing the service attempting to start before all the programs have been defined.

```bash
Notice: /Stage[main]/Supervisord/Supervisord::Group[my-workers]/File[/etc/supervisor.d/group_my-workers.conf]/ensure: created
Debug: /Stage[main]/Supervisord/Supervisord::Group[my-workers]/File[/etc/supervisor.d/group_my-workers.conf]: The container Supervisord::Group[my-workers] will propagate my refresh event
Info: /Stage[main]/Supervisord/Supervisord::Group[my-workers]/File[/etc/supervisor.d/group_my-workers.conf]: Scheduling refresh of Class[Supervisord::Reload]
Debug: Supervisord::Group[my-workers]: The container Class[Supervisord] will propagate my refresh event
Notice: /Stage[main]/Supervisord::Config/Concat::Fragment[supervisord_unix]/File[/var/lib/puppet/concat/_etc_supervisord.conf/fragments/01_supervisord_unix]/ensure: defined content as '{md5}2a371bf6742698a806a2ab9879d60540'
Info: /Stage[main]/Supervisord::Config/Concat::Fragment[supervisord_unix]/File[/var/lib/puppet/concat/_etc_supervisord.conf/fragments/01_supervisord_unix]: Scheduling refresh of Exec[concat_/etc/supervisord.conf]
Debug: /Stage[main]/Supervisord::Config/Concat::Fragment[supervisord_unix]/File[/var/lib/puppet/concat/_etc_supervisord.conf/fragments/01_supervisord_unix]: The container Concat::Fragment[supervisord_unix] will propagate my refresh event
Debug: Exec[concat_/etc/supervisord.conf](provider=posix): Executing check '/var/lib/puppet/concat/bin/concatfragments.sh -o "/var/lib/puppet/concat/_etc_supervisord.conf/fragments.concat.out" -d "/var/lib/puppet/concat/_etc_supervisord.conf" -t'
Debug: Executing '/var/lib/puppet/concat/bin/concatfragments.sh -o "/var/lib/puppet/concat/_etc_supervisord.conf/fragments.concat.out" -d "/var/lib/puppet/concat/_etc_supervisord.conf" -t'
Debug: /Stage[main]/Supervisord::Config/Concat[/etc/supervisord.conf]/Exec[concat_/etc/supervisord.conf]/unless: cmp: EOF on /var/lib/puppet/concat/_etc_supervisord.conf/fragments.concat.out
Debug: Exec[concat_/etc/supervisord.conf](provider=posix): Executing '/var/lib/puppet/concat/bin/concatfragments.sh -o "/var/lib/puppet/concat/_etc_supervisord.conf/fragments.concat.out" -d "/var/lib/puppet/concat/_etc_supervisord.conf"'
Debug: Executing '/var/lib/puppet/concat/bin/concatfragments.sh -o "/var/lib/puppet/concat/_etc_supervisord.conf/fragments.concat.out" -d "/var/lib/puppet/concat/_etc_supervisord.conf"'
Notice: /Stage[main]/Supervisord::Config/Concat[/etc/supervisord.conf]/Exec[concat_/etc/supervisord.conf]/returns: executed successfully
Debug: /Stage[main]/Supervisord::Config/Concat[/etc/supervisord.conf]/Exec[concat_/etc/supervisord.conf]: The container Concat[/etc/supervisord.conf] will propagate my refresh event
Debug: Exec[concat_/etc/supervisord.conf](provider=posix): Executing check '/var/lib/puppet/concat/bin/concatfragments.sh -o "/var/lib/puppet/concat/_etc_supervisord.conf/fragments.concat.out" -d "/var/lib/puppet/concat/_etc_supervisord.conf" -t'
Debug: Executing '/var/lib/puppet/concat/bin/concatfragments.sh -o "/var/lib/puppet/concat/_etc_supervisord.conf/fragments.concat.out" -d "/var/lib/puppet/concat/_etc_supervisord.conf" -t'
Notice: /Stage[main]/Supervisord::Config/Concat[/etc/supervisord.conf]/Exec[concat_/etc/supervisord.conf]: Triggered 'refresh' from 4 events
Debug: /Stage[main]/Supervisord::Config/Concat[/etc/supervisord.conf]/Exec[concat_/etc/supervisord.conf]: The container Concat[/etc/supervisord.conf] will propagate my refresh event
Notice: /Stage[main]/Supervisord::Config/Concat[/etc/supervisord.conf]/File[/etc/supervisord.conf]/ensure: defined content as '{md5}2712bb67930fa94c5ab40515280dbe33'
Debug: /Stage[main]/Supervisord::Config/Concat[/etc/supervisord.conf]/File[/etc/supervisord.conf]: The container Concat[/etc/supervisord.conf] will propagate my refresh event
Debug: Concat::Fragment[supervisord_unix]: The container Class[Supervisord::Config] will propagate my refresh event
Debug: Concat[/etc/supervisord.conf]: The container Class[Supervisord::Config] will propagate my refresh event
Debug: Class[Supervisord::Config]: The container Stage[main] will propagate my refresh event
Info: Class[Supervisord::Config]: Scheduling refresh of Class[Supervisord::Reload]
Info: Class[Supervisord::Service]: Scheduling refresh of Service[supervisord]
Debug: Service[supervisord](provider=upstart): Could not find supervisord.conf in /etc/init
Debug: Service[supervisord](provider=upstart): Could not find supervisord.conf in /etc/init.d
Debug: Service[supervisord](provider=upstart): Could not find supervisord in /etc/init
Debug: Executing '/etc/init.d/supervisord status'
Debug: Executing '/etc/init.d/supervisord start'
Error: Could not start Service[supervisord]: Execution of '/etc/init.d/supervisord start' returned 2: 
Error: /Stage[main]/Supervisord::Service/Service[supervisord]/ensure: change from stopped to running failed: Could not start Service[supervisord]: Execution of '/etc/init.d/supervisord start' returned 2: 
Debug: Executing '/etc/init.d/supervisord status'
Debug: /Stage[main]/Supervisord::Service/Service[supervisord]: Skipping restart; service is not running
Notice: /Stage[main]/Supervisord::Service/Service[supervisord]: Triggered 'refresh' from 1 events
Debug: /Stage[main]/Supervisord::Service/Service[supervisord]: The container Class[Supervisord::Service] will propagate my refresh event
Debug: Class[Supervisord::Service]: The container Stage[main] will propagate my refresh event
Notice: /Stage[main]/Supervisord/Anchor[supervisord::end]: Dependency Service[supervisord] has failures: true
Warning: /Stage[main]/Supervisord/Anchor[supervisord::end]: Skipping because of failed dependencies
Notice: /Stage[main]/Supervisord/Supervisord::Program[my-worker-search-consumer]/File[/etc/supervisor.d/program_my-worker-search-consumer.conf]/ensure: created
Debug: /Stage[main]/Supervisord/Supervisord::Program[my-worker-search-consumer]/File[/etc/supervisor.d/program_my-worker-search-consumer.conf]: The container Supervisord::Program[my-worker-search-consumer] will propagate my refresh event
Info: /Stage[main]/Supervisord/Supervisord::Program[my-worker-search-consumer]/File[/etc/supervisor.d/program_my-worker-search-consumer.conf]: Scheduling refresh of Class[Supervisord::Reload]
Debug: Supervisord::Program[my-worker-search-consumer]: The container Class[Supervisord] will propagate my refresh event
```

Adding the following to my manifest corrected the issue for me, but I'm sure this should be handled automatically.

```puppet
# Ensure proper supervisord ordering
Class['supervisord::config']
-> Supervisord::Program <| |>
-> Supervisord::Group <| |>
```

I would gladly submit a PR for this fix, if someone can enlighten me as to where these resources should be anchored.

**Comment by ajcrowe** [comment]
Not seen that one before, but your logic is correct if config has not created all the appropriate folders it would cause issues. I'll look at adding this additional ordering to `init.pp`

**Comment by ajcrowe** [comment]
I have committed a fix in 7681e53e for this and will be in the next release.

Thanks for the contribution

**Comment by mcrumm** [comment]
Awesome, thank you!
